src/Compilers/Core/Portable/SourceGeneration/Nodes/IncrementalValueSources.cs

```diff
@@ -99,7 +77,7 @@ public ArrayBuilder<ISyntaxTransformNode> SyntaxTransformNodes
                         // but the rest of the structure isn't. We should decide if we want to support that or 
                         // just say this whole thing isn't thread safe.
 
-                        var newNodes = ArrayBuilder<ISyntaxTransformNode>.GetInstance();
+                        var newNodes = ArrayBuilder<ISyntaxInputNode>.GetInstance();
                         InterlockedOperations.Initialize(ref _transformNodes, newNodes);
```

Comment:
>InterlockedOperations.Initialize [](http://example.com/codeflow?start=24&length=32)

Could we simplify this by initializing `_transformNodes` in the constructor? (Won't most uses of `Builder` add at least one node?)

Reply:
Yeah I think thats probably a sensible step at this point. Its only generators that look at syntax, but thats probably a large enough percentage of them to just assume. If we find in profiling that this is causing enough issues we can restore something along these lines transparently to the public api.

---
In reply to: [632130244](https://github.com/dotnet/roslyn/pull/53383#discussion_r632130244) [](ancestors = 632130244)

Reply:
Hmm, turns out that now we have an interface for the ISyntaxInputNodes, we can just replace this whole type with an array builder. Much cleaner!

---
In reply to: [632152657](https://github.com/dotnet/roslyn/pull/53383#discussion_r632152657) [](ancestors = 632152657,632130244)